src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+#include "XtensaOptimize.h"
+
+#include "AlignLoads.h"
+#include "Bounds.h"
+#include "CSE.h"
+#include "ConciseCasts.h"
+#include "Expr.h"
+#include "ExprUsesVar.h"
+#include "FindIntrinsics.h"
+#include "IREquality.h"
+#include "IRMatch.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Lerp.h"
+#include "LoopCarry.h"
+#include "Simplify.h"
+#include "Substitute.h"
+#include <utility>
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+
+using namespace Halide::ConciseCasts;
+
+template<>
+bool is_native_xtensa_vector<int8_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<int8_t>();
+    return t.is_int() && (t.bits() == 8) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<uint8_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<uint8_t>();
+    return t.is_uint() && (t.bits() == 8) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<int16_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<int16_t>();
+    return t.is_int() && (t.bits() == 16) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<uint16_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<uint16_t>();
+    return t.is_uint() && (t.bits() == 16) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<int32_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<int32_t>();
+    return t.is_int() && (t.bits() == 32) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<int64_t>(const Type &t, const Target &target) {
+    // On Xtensa int64 vectors are *wide* vectors, so the number of lanes match
+    // the number of lanes for 32-bit vectors.
+    int vector_size = target.natural_vector_size<int32_t>();
+    return t.is_int() && (t.bits() == 64) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<uint32_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<uint32_t>();
+    return t.is_uint() && (t.bits() == 32) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<float16_t>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<float16_t>();
+    return t.is_float() && (t.bits() == 16) && (t.lanes() == vector_size);
+}
+
+template<>
+bool is_native_xtensa_vector<float>(const Type &t, const Target &target) {
+    int vector_size = target.natural_vector_size<float>();
+    return t.is_float() && (t.bits() == 32) && (t.lanes() == vector_size);
+}
+
+bool is_native_vector_type(const Type &t, const Target &target) {
```

Comment:
Is there an explanation as to why 8 and 24 bit are full lanes, 16 and 48 are half lanes, and 32 bits are a fourth of the 8 bit lanes? This behavior makes very little sense to me, but I am not familiar with this target.

Reply:
e.g. I don't understand why this method isn't as simple as `is_double_native_vector_type` below.